documentation/interrupt_handling.md

```diff
@@ -109,15 +189,67 @@ When the monitor gets an FIQ exception it:
 6. FIQ is received as in exception using the state vector
-7. Secure world issues an SMC to return to normal world
-8. Monitor saves secure world context and restores normal world context
-9. Does a return from exception into restored context
+7. The state vector handle returns from exception in secure world
+8. Secure world issues an SMC to return to normal world
+9. Monitor saves secure world context and restores normal world context
+10. Does a return from exception into restored context
 
 ![FIQ received when SCR_NS is set](images/interrupt_handling/fiq.png "FIQ received when SCR_NS is set")
 
 ![FIQ received while processing an IRQ forwarded from secure world](images/interrupt_handling/irq_fiq.png "FIQ received while processing an IRQ forwarded from secure world")
 
-## Deliver FIQ to secure world when SCR_NS is cleared
+## 4.2. Deliver FIQ to secure world when SCR_NS is cleared
+
 Since SCR_FIQ is cleared when SCR_NS is cleared a FIQ will be delivered
 using the state vector (VBAR) in secure world. The FIQ is received as any
 other exception by Trusted OS, the monitor is not involved at all.
+
+# 5. TEE Thread Scheduling
+
+## 5.1. TEE Thread
+When the normal world invokes the optee_os, the Monitor is in charge of saving
+and restoring caller and optee_os contextes while optee_os is in charge of
+assigning a TEE thread for the excution of the invoked service. The TEE thread
+terminates when optee_os returns to the normal world with a service completion
+status.
+
+A TEE thread execution can be interrutped by a native interrupt. In this case
+the native interrupt is handled from the interrupt exception handlers and once
+served, optee_os returns to the execution TEE thread.
+
+A TEE thread execution can be interrutped by a foreign interrupt. In this
+case, optee_os suspends the TEE thread and invokes the normal world through the
+Monitor (optee_os so-called RPC services). Resuming the TEE threads requires
+from the normal world to inovke optee_os with the expected RPC service status.
+
+A TEE thread execution can lead optee_os in invoking a service in normal
+world: access a file, get the REE current time, ...
+
+## 5.2. Scheduling Considerations
+
+When a TEE thread is interrupted by a foreign interrupt and when optee_os
+invokes a normal world service, the normal world gets the opportunity to
+reschedule the running applications. Since the TEE thread will resume only
+once the client application is scheduled back. Thus, a TEE thread execution
+follows the scheduling of the normal world caller context.
+
+Optee_os does not implement any thread scheduling. Each TEE thread is expected
+to track a service that is invoked from the normal world and should return to
+it with a execution status.
+
+The proposed OP-TEE Linux driver (as implemented in drivers/tee/optee since
+Linux kernel 4.12) is designed so that the Linux thread invoking OP-TEE gets
+assigned an OP-TEE thread on TEE side. The execution of the OP-TEE thread is
```

Comment:
> Linux thread invoking OP-TEE gets assigned an OP-TEE thread on TEE side.

Does it applies for fast SMC calls also?

Reply:
no, only to standard SMCs. I will fix.